### PR TITLE
Allow generating presigned URIs for S3 and GCS

### PR DESCRIPTION
**Pull request by wendigo**
## Release notes

( ) This is not user-visible or is docs only, and no release notes are required.
( ) Release notes are required. Please propose a release note for me.
( ) Release notes are required, with the following suggested text:

```markdown
# Section
* Fix some things. ({issue}`issuenumber`)
```

### CODE REVIEW

File: lib/trino-filesystem/src/main/java/io/trino/filesystem/TrinoFileSystem.java [hunk 1]
```diff
@@ -228,6 +229,18 @@ Set<Location> listDirectories(Location location)
...
+    /**
+     * Returns the direct pre-signed URI location for the given storage location.
+     *
+     * This is used to retrieve the file directly from the storage location without going through
+     * the Trino server. This is useful for large files where the Trino server would be a bottleneck.
+     */
+    default Optional<UriLocation> preSignedUri(Location location, Duration ttl)
+            throws IOException
+    {
+        return Optional.empty();
```
Comment (losipiuk) [inline lib/trino-filesystem/src/main/java/io/trino/filesystem/TrinoFileSystem.java:241]: What does return of Optional.empty() mean? Please add info to the Javadoc. 
Is it "not supported" flag? I would prefer to throw exception in that case, and have some separate method from which you may get filesystem "capabilities".

File: lib/trino-filesystem/src/test/java/io/trino/filesystem/AbstractTestTrinoFileSystem.java [hunk 2]
```diff
@@ -950,6 +962,69 @@ public void testListFiles()
         testListFiles(isHierarchical());
     }
 
+    @Test
+    public void testPreSignedUris()
+            throws IOException
+    {
+        if (!supportsPreSignedUri()) {
+            return;
```
Comment (losipiuk) [inline lib/trino-filesystem/src/test/java/io/trino/filesystem/AbstractTestTrinoFileSystem.java:970]: `abort("pre signed urls not supported by %s"....)

File: lib/trino-filesystem/src/test/java/io/trino/filesystem/AbstractTestTrinoFileSystem.java [hunk 3]
```diff
@@ -950,6 +962,69 @@ public void testListFiles()
...
+            assertThat(directLocation).isPresent();
+            assertThat(retrieveUri(directLocation.get()))
+                    .isEqualTo(TEST_BLOB_CONTENT_PREFIX + location);
+
+            // Check if can be retrieved more than once
+            assertThat(retrieveUri(directLocation.get()))
+                    .isEqualTo(TEST_BLOB_CONTENT_PREFIX + location);
+
+            // Check if after a timeout the link is no longer valid
+            assertEventually(() -> assertThatThrownBy(() -> retrieveUri(directLocation.get()))
```
Comment (losipiuk) [inline lib/trino-filesystem/src/test/java/io/trino/filesystem/AbstractTestTrinoFileSystem.java:987]: nit: make timeout shorted - default is 30s

File: lib/trino-filesystem/src/test/java/io/trino/filesystem/AbstractTestTrinoFileSystem.java [hunk 4]
```diff
@@ -950,6 +962,69 @@ public void testListFiles()
...
+        }
+        catch (InterruptedException e) {
+            Thread.currentThread().interrupt();
+            throw new RuntimeException(e);
+        }
+    }
+
+    private static HttpRequest.Builder addHeaders(HttpRequest.Builder builder, Map<String, List<String>> headers)
+    {
+        for (Map.Entry<String, List<String>> entry : headers.entrySet()) {
```
Comment (losipiuk) [inline lib/trino-filesystem/src/test/java/io/trino/filesystem/AbstractTestTrinoFileSystem.java:1016]:
```
headers.forEach( (k, v) -> {...
```

File: lib/trino-filesystem-gcs/src/main/java/io/trino/filesystem/gcs/GcsFileSystem.java [hunk 5]
```diff
@@ -272,6 +280,23 @@ public Optional<Location> createTemporaryDirectory(Location targetPath, String t
...
     }
 
+    @Override
+    public Optional<UriLocation> preSignedUri(Location location, Duration ttl)
+            throws IOException
+    {
+        GcsLocation gcsLocation = new GcsLocation(location);
+        BlobInfo blobInfo = BlobInfo.newBuilder(BlobId.of(gcsLocation.bucket(), gcsLocation.path())).build();
+
+        URL url = storage.signUrl(blobInfo, ttl.toMillis(), TimeUnit.MILLISECONDS, withV4Signature());
```
Comment (losipiuk) [inline lib/trino-filesystem-gcs/src/main/java/io/trino/filesystem/gcs/GcsFileSystem.java:290]: static import MILLISECONDS

**Review by losipiuk — COMMENTED** [review]
lgtm

File: lib/trino-filesystem-s3/src/main/java/io/trino/filesystem/s3/S3FileSystemLoader.java [hunk 6]
```diff
@@ -167,6 +177,60 @@ else if (iamRole.isPresent()) {
...
+        Optional<AwsCredentialsProvider> staticCredentialsProvider = createStaticCredentialsProvider(config);
+        Optional<String> staticRegion = Optional.ofNullable(config.getRegion());
+        Optional<String> staticEndpoint = Optional.ofNullable(config.getEndpoint());
+        boolean pathStyleAccess = config.isPathStyleAccess();
+        boolean useWebIdentityTokenCredentialsProvider = config.isUseWebIdentityTokenCredentialsProvider();
+        Optional<String> staticIamRole = Optional.ofNullable(config.getIamRole());
+        String staticRoleSessionName = config.getRoleSessionName();
+        String externalId = config.getExternalId();
+
+        return mapping -> {
```
Comment (electrum) [inline lib/trino-filesystem-s3/src/main/java/io/trino/filesystem/s3/S3FileSystemLoader.java:191]: We don't need to support S3 security mapping for this, so we can simplify this code.

**Review by electrum — APPROVED** [review]
I like how the GCS code is so much simpler than S3 due to not needing an entirely different API and entirely different `Builder`

File: lib/trino-filesystem-s3/src/main/java/io/trino/filesystem/s3/S3FileSystemLoader.java (hunk 6 above)
Comment (wendigo) [reply lib/trino-filesystem-s3/src/main/java/io/trino/filesystem/s3/S3FileSystemLoader.java:191]: Ok

**Comment by wendigo** [comment]
@electrum I can't get azure to work. I will ask someone from Microsoft for help and move forward without it

**Comment by colebow** [comment]
Release notes needed?

**Comment by wendigo** [comment]
@colebow no

**Pull request merged**